src/Processors/Transforms/Streaming/JoinTransformWithAlignment.cpp

```diff
@@ -24,113 +31,78 @@ JoinTransformWithAlignment::JoinTransformWithAlignment(
+    right_input.watermark_column_type = right_join_stream_desc->alignmentKeyColumnType();
 
-    left_watermark_column_position = left_join_stream_desc->lagBehindColumnPosition();
-    left_watermark_column_type = left_join_stream_desc->lagBehindColumnType();
+    left_input.is_changelog_input = isChangelogDataStream(left_join_stream_desc->data_stream_semantic);
+    right_input.is_changelog_input = isChangelogDataStream(right_join_stream_desc->data_stream_semantic);
 
-    right_watermark_column_position = right_join_stream_desc->lagBehindColumnPosition();
-    right_watermark_column_type = right_join_stream_desc->lagBehindColumnType();
+    /// Some join combinations have builtin buffer to do aligning, so don't build an extra external buffer to align:
+    /// 1) For asof join, the right side has builtin buffer to keep multiple versions
+    /// 2) For range join, the left and right side both have builtin buffer to keep multiple ranges
+    left_input.need_aligned_buffer = !join->leftHasBuiltInAlignedBuffer();
+    right_input.need_aligned_buffer = !join->rightHasBuiltInAlignedBuffer();
 
     assert(
-        (left_watermark_column_position && right_watermark_column_position)
-        || (!left_watermark_column_position && !right_watermark_column_position));
+        (left_input.watermark_column_position && right_input.watermark_column_position)
+        || (!left_input.watermark_column_position && !right_input.watermark_column_position));
+
+    assert(
+        (left_input.watermark_column_type && right_input.watermark_column_type)
+        || (!left_input.watermark_column_type && !right_input.watermark_column_type));
+
+    latency_threshold = left_join_stream_desc->latency_threshold;
+
+    /// When left stream or right stream is in quiesce, we still need to push execute the join probably
+    quiesce_threshold_ms = left_join_stream_desc->quiesce_threshold_ms;
+    if (quiesce_threshold_ms <= 0)
+        quiesce_threshold_ms = std::max<Int64>(std::abs(latency_threshold) / 2, 500);
 
-    assert((left_watermark_column_type && right_watermark_column_type) || (!left_watermark_column_type && !right_watermark_column_type));
+    left_input.last_data_ts = right_input.last_data_ts = DB::MonotonicMilliseconds::now();
 
     LOG_INFO(
         log,
-        "quiesce_threshold_ms={} lag_latency_threshold={} left_lag_column={}, right_lag_column={}",
+        "quiesce_threshold_ms={} lag_latency_threshold={} left_alignment_column={}, right_alignment_column={}",
         quiesce_threshold_ms,
         latency_threshold,
-        left_join_stream_desc->lag_column,
-        right_join_stream_desc->lag_column);
+        left_join_stream_desc->alignment_column,
+        right_join_stream_desc->alignment_column);
 }
 
-IProcessor::Status JoinTransformWithAlignment::prepareRightInput()
+IProcessor::Status JoinTransformWithAlignment::prepareInput(InputPortWithData & input_with_data)
```

Comment:
After we refactor out different modes, the prepareInput is asymmetrix for asof join.